File: src/main/converters/proprietary/ncmDecoder.ts
import { createDecipheriv } from 'crypto'
import { readFile } from 'fs/promises'

const ncmMagic = Buffer.from([0x43, 0x54, 0x45, 0x4e, 0x46, 0x44, 0x41, 0x4d])
const coreKey = Buffer.from('hzHRAmso5kInbaxW', 'utf8')

function decryptAesEcb(data: Buffer, key: Buffer): Buffer {
  const decipher = createDecipheriv('aes-128-ecb', key, null)
  decipher.setAutoPadding(true)
  return Buffer.concat([decipher.update(data), decipher.final()])
}

function buildKeyBox(keyData: Buffer): Uint8Array {
  const keyBox = new Uint8Array(256)

  for (let i = 0; i < 256; i += 1) keyBox[i] = i

  let lastByte = 0
  let keyOffset = 0

  for (let i = 0; i < 256; i += 1) {
    const swap = keyBox[i]
    const c = (swap + lastByte + keyData[keyOffset]) & 0xff
    keyOffset = (keyOffset + 1) % keyData.length
    keyBox[i] = keyBox[c]
    keyBox[c] = swap
    lastByte = c
  }

  return keyBox
}

function applyNcmKeyStream(data: Buffer, keyBox: Uint8Array): Buffer {
  const output = Buffer.from(data)

  for (let i = 0; i < output.length; i += 1) {
    const j = (i + 1) & 0xff
    output[i] ^= keyBox[(keyBox[j] + keyBox[(keyBox[j] + j) & 0xff]) & 0xff]
  }

  return output
}

export async function decodeNcmContainer(inputPath: string): Promise<Buffer> {
  const buffer = await readFile(inputPath)

  if (!buffer.subarray(0, 8).equals(ncmMagic)) {
    throw new Error('NCM 魔数校验失败')
  }

  let offset = 10
  const keyLength = buffer.readUInt32LE(offset)
  offset += 4

  const encryptedKey = Buffer.from(buffer.subarray(offset, offset + keyLength))
  offset += keyLength

  for (let i = 0; i < encryptedKey.length; i += 1) encryptedKey[i] ^= 0x64

  const decryptedKey = decryptAesEcb(encryptedKey, coreKey).subarray(17)
  if (decryptedKey.length === 0) throw new Error('NCM 音频密钥为空')

  const keyBox = buildKeyBox(decryptedKey)
  const metadataLength = buffer.readUInt32LE(offset)
  offset += 4 + metadataLength

  const crcAndGapLength = 9
  offset += crcAndGapLength

  const imageSize = buffer.readUInt32LE(offset)
  offset += 4 + imageSize

  if (offset >= buffer.length) throw new Error('NCM 音频负荷为空')

  return applyNcmKeyStream(buffer.subarray(offset), keyBox)
}

File: android-apk/app/src/main/java/com/volcanic/musicplayer/decoder/NcmDecoder.java
package com.volcanic.musicplayer.decoder;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;

final class NcmDecoder {
    private static final byte[] NCM_MAGIC = new byte[]{
            0x43, 0x54, 0x45, 0x4e, 0x46, 0x44, 0x41, 0x4d
    };
    private static final byte[] CORE_KEY = "hzHRAmso5kInbaxW".getBytes(StandardCharsets.UTF_8);

    private NcmDecoder() {
    }

    static byte[] decode(File inputFile) throws Exception {
        byte[] input = PrivateContainerDecoder.readAll(inputFile);
        if (input.length < 32 || !startsWith(input, NCM_MAGIC)) {
            throw new IllegalArgumentException("Invalid NCM header");
        }

        int offset = 10;
        int keyLength = readUInt32LE(input, offset);
        offset += 4;
        if (keyLength <= 0 || offset + keyLength > input.length) {
            throw new IllegalArgumentException("Invalid NCM key block");
        }

        byte[] encryptedKey = Arrays.copyOfRange(input, offset, offset + keyLength);
        offset += keyLength;
        for (int i = 0; i < encryptedKey.length; i++) {
            encryptedKey[i] = (byte) ((encryptedKey[i] & 0xff) ^ 0x64);
        }

        byte[] decryptedKey = aesEcbDecrypt(encryptedKey, CORE_KEY);
        if (decryptedKey.length <= 17) {
            throw new IllegalArgumentException("Empty NCM audio key");
        }
        byte[] keyData = Arrays.copyOfRange(decryptedKey, 17, decryptedKey.length);
        byte[] keyBox = buildKeyBox(keyData);

        int metadataLength = readUInt32LE(input, offset);
        offset += 4 + metadataLength;
        offset += 9;
        int imageSize = readUInt32LE(input, offset);
        offset += 4 + imageSize;
        if (offset >= input.length) {
            throw new IllegalArgumentException("Empty NCM payload");
        }

        byte[] payload = Arrays.copyOfRange(input, offset, input.length);
        applyKeyStream(payload, keyBox);
        return payload;
    }

    private static byte[] aesEcbDecrypt(byte[] encrypted, byte[] key) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
        cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, "AES"));
        return cipher.doFinal(encrypted);
    }

    private static byte[] buildKeyBox(byte[] keyData) {
        byte[] keyBox = new byte[256];
        for (int i = 0; i < 256; i++) {
            keyBox[i] = (byte) i;
        }

        int lastByte = 0;
        int keyOffset = 0;
        for (int i = 0; i < 256; i++) {
            int swap = keyBox[i] & 0xff;
            int c = (swap + lastByte + (keyData[keyOffset] & 0xff)) & 0xff;
            keyOffset = (keyOffset + 1) % keyData.length;
            keyBox[i] = keyBox[c];
            keyBox[c] = (byte) swap;
            lastByte = c;
        }
        return keyBox;
    }

    private static void applyKeyStream(byte[] payload, byte[] keyBox) {
        for (int i = 0; i < payload.length; i++) {
            int j = (i + 1) & 0xff;
            int index = ((keyBox[j] & 0xff) + j) & 0xff;
            int maskIndex = ((keyBox[j] & 0xff) + (keyBox[index] & 0xff)) & 0xff;
            payload[i] = (byte) ((payload[i] & 0xff) ^ (keyBox[maskIndex] & 0xff));
        }
    }

    private static int readUInt32LE(byte[] data, int offset) {
        if (offset < 0 || offset + 4 > data.length) {
            throw new IllegalArgumentException("Unexpected NCM EOF");
        }
        return ByteBuffer.wrap(data, offset, 4).order(ByteOrder.LITTLE_ENDIAN).getInt();
    }

    private static boolean startsWith(byte[] input, byte[] prefix) {
        if (input.length < prefix.length) {
            return false;
        }
        for (int i = 0; i < prefix.length; i++) {
            if (input[i] != prefix[i]) {
                return false;
            }
        }
        return true;
    }
}
